src/interface/trsm.hpp

```diff
@@ -0,0 +1,432 @@
+        auto gemmBufferA = invA + gemmOffsetA;
+        auto gemmBufferB = B + gemmOffsetB;
+        auto gemmBufferC = X + gemmOffsetC;
+        auto gemmEvent =
+            internal::_gemm(ex, gemmTransA, gemmTransB, gemmM, gemmN, gemmK,
+                            gemmAlpha, gemmBufferA, gemmLda, gemmBufferB,
+                            gemmLdb, gemmBeta, gemmBufferC, gemmLdc);
+        trsmEvents = concatenate_vectors(trsmEvents, gemmEvent);
+
+        if ((i - blockSize) < 0) {
+          break;
+        }
+
+        gemmM = i;
+        gemmN = N;
+        gemmK = currentBlockSize;
+        gemmTransA = isTranspose ? 't' : 'n';
+        gemmTransB = 'n';
+        gemmAlpha = element_t{-1};
+        gemmBeta = (i == iStart) ? alpha : element_t{1};
+        gemmLda = lda;
+        gemmLdb = ldx;
+        gemmLdc = ldb;
+        gemmOffsetA = !isTranspose ? (i * lda) : i;
+        gemmOffsetB = i;
+        gemmOffsetC = 0;
+        gemmBufferA = A + gemmOffsetA;
+        gemmBufferB = X + gemmOffsetB;
+        gemmBufferC = B + gemmOffsetC;
+        gemmEvent =
+            internal::_gemm(ex, gemmTransA, gemmTransB, gemmM, gemmN, gemmK,
+                            gemmAlpha, gemmBufferA, gemmLda, gemmBufferB,
+                            gemmLdb, gemmBeta, gemmBufferC, gemmLdc);
+        trsmEvents = concatenate_vectors(trsmEvents, gemmEvent);
+      }
+    }
+  } else {
+    // Right side
+
+    if ((isUpper && isTranspose) || (!isUpper && !isTranspose)) {
+      // Solves the system XA = alpha*B when A is lower triangular
+
+      // True when (lower triangular) or (upper triangular and transposed)
+      const index_t specialBlockSize =
+          (N % blockSize == 0) ? blockSize : (N % blockSize);
+      const index_t iStart = N - specialBlockSize;
+      for (int i = iStart; i >= 0; i -= blockSize) {
+        const index_t currentBlockSize =
+            (i == iStart) ? specialBlockSize : blockSize;
+
+        index_t gemmM = M;
+        index_t gemmN = currentBlockSize;
+        index_t gemmK = currentBlockSize;
+        char gemmTransA = 'n';
+        char gemmTransB = isTranspose ? 't' : 'n';
+        element_t gemmAlpha = (i == iStart) ? alpha : element_t{1};
+        element_t gemmBeta = element_t{0};
+        index_t gemmLda = ldb;
+        index_t gemmLdb = blockSize;
+        index_t gemmLdc = ldx;
+        std::ptrdiff_t gemmOffsetA = i * ldb;
+        std::ptrdiff_t gemmOffsetB = i * blockSize;
+        std::ptrdiff_t gemmOffsetC = i * ldx;
+        auto gemmBufferA = B + gemmOffsetA;
+        auto gemmBufferB = invA + gemmOffsetB;
+        auto gemmBufferC = X + gemmOffsetC;
+        auto gemmEvent =
+            internal::_gemm(ex, gemmTransA, gemmTransB, gemmM, gemmN, gemmK,
+                            gemmAlpha, gemmBufferA, gemmLda, gemmBufferB,
+                            gemmLdb, gemmBeta, gemmBufferC, gemmLdc);
+        trsmEvents = concatenate_vectors(trsmEvents, gemmEvent);
+
```

Comment:
Same as above these operations are repeated multiple times. Can you wrap them in a function and call the function instead?